internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package loki
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/cmd"
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const maxRequestEntries = 100
+
+var requestTimeout = 10 * time.Second
+
+type Client struct {
+	targetName string
+	remoteURL  string
+	entries    []lokiEntryWithService
+
+	httpClient *http.Client
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return &Client{
+		targetName: target.Name,
+		remoteURL:  target.Location,
+		httpClient: &http.Client{Timeout: requestTimeout},
+	}
+}
+
+func (c *Client) AddLog(entry servicelog.Entry) error {
+	c.entries = append(c.entries, lokiEntryWithService{
+		entry:   encodeEntry(entry),
+		service: entry.Service,
+	})
+	c.truncateBuffer()
```

Comment:
Instead of truncate buffer here I would do the following due to the memory leak by truncating the front of the slice (as it can't free the underlying array).

```
const reallocBufferThreshold = maxRequestEntries*2
func (c *Client) AddLog(entry servicelog.Entry) error {
	if l := len(c.entries); l > maxRequestEntries {
		c.entries = c.entries[(l - maxRequestEntries):]
	}
	if cap(c.entries) == 0 {
		c.entries = append(make([]lokiEntryWithService, 0, reallocBufferThreshold), c.entries...)
	}
	c.entries = append(c.entries, lokiEntryWithService{
		entry:   encodeEntry(entry),
		service: entry.Service,
	})
	return nil
}
```

Reply:
edited a few times, I think that's what I'm trying to go for.

Essentially we don't want to reallocate the underlying array too frequently. This way we use a bit more memory, but don't need to implement a ring buffer.